Generalise getInitialData Logging

ryanmccombe opened the pull request:
Resolves #NUMBER

**Overall change:**
Refactor On-Demand Radio's `getInitialData` logging

Updated to a more generalised pattern that could also be used for other pages

**Code changes:**

- _Update `pathWithLogging` location and API._
- _Refactor On-Demand Radio getInitialData._

---

- [ ] I have assigned myself to this PR and the corresponding issues
- [ ] I have added labels to this PR for the relevant pod(s) affected by these changes
- [ ] I have assigned this PR to the Simorgh project

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [ ] This PR requires manual testing

ryanmccombe committed:
Refactor OD-radio getInitialData

ryanmccombe committed:
Merge branch 'latest' into refactor-od-get-initial-data

karinathomasbbc commented on src/app/lib/utilities/logging/pathWithLogging/index.js:
```diff
@@ -19,15 +18,20 @@ const getLoggingFunction = level =>
   }[level] || logger.info);
 
-export default (fieldPath, { logLevel = LOG_LEVELS.INFO } = {}) => pageData => {
+const pathWithLogging = (url, logCategory, pageData) => (
```
Can we maybe add some docs here to describe what this function is doing?

ryanmccombe replied:
Added readme.md

jroebu14 reviewed (commented):
Looks good. Wondering if a better API would be for `get` to have a second argument called `logUndefined` or something.

What happens if we get no data back. Will there be about a dozen logs?

If the brandTitle comes back undefined and we just log it but don't throw an error, is that the best way? Is a better handler to have required fields that the page data is matched against and if not met then an error is thrown and logged in a catch?

rebeccamcginn committed:
Merge branch 'latest' into refactor-od-get-initial-data

rebeccamcginn committed:
Merge branch 'latest' into refactor-od-get-initial-data

ryanmccombe committed:
CR feedback

ryanmccombe commented:
> Looks good. Wondering if a better API would be for get to have a second argument called logUndefined or something.

👍 Updated it to be more like this - reduces the code in `getInitialData` a good bit

> What happens if we get no data back. Will there be about a dozen logs?

When there's no data, it looks like we crash (from the infamous `.split`) before any logging from this function happens

I've just updated the function to return early if we get no data back: https://github.com/bbc/simorgh/blob/05be47f822fbc3549c83f8f19d91619b39dcffd9/src/app/routes/onDemandRadio/getInitialData/index.js#L40-L44

This also makes it fail more gracefully (shows 404 page instead of complaining about `.split`)

> If the brandTitle comes back undefined and we just log it but don't throw an error, is that the best way? Is a better handler to have required fields that the page data is matched against and if not met then an error is thrown and logged in a catch?

Not sure - it is how we have tended to do it on other pages - when we have janky data, we still make a "best attempt" at rendering.  Curious what others / product think, though

ryanmccombe committed:
Merge branch 'latest' into refactor-od-get-initial-data

ryanmccombe committed:
Improve readme

ryanmccombe committed:
Merge branch 'refactor-od-get-initial-data' of github.com:bbc/simorgh into refactor-od-get-initial-data

ryanmccombe committed:
Fix logic error

jroebu14 commented:
Looks good to me. Just a question, is certain fields missing from data a common problem we have?

rebeccamcginn commented:
> Not sure - it is how we have tended to do it on other pages - when we have janky data, we still make a "best attempt" at rendering. Curious what others / product think, though

The media is still playable. We have scenarios when we don't render the media player which is worse than not rendering the brand title. The brand title should not be missing from the Ares response so I think this is unlikely to happen.

rebeccamcginn committed:
Merge branch 'latest' into refactor-od-get-initial-data

rebeccamcginn committed:
Merge branch 'latest' into refactor-od-get-initial-data

jroebu14 commented on src/app/routes/onDemandRadio/getInitialData/index.js:
```diff
@@ -62,26 +32,17 @@ const getEpisodeAvailableUntil = path([
   const onDemandRadioDataPath = overrideRendererOnTest(pathname);
   const { json, ...rest } = await fetchPageData(onDemandRadioDataPath);
+
+  // If no data was returned by fetchPageData, we can just skip all the data processing
```
I understand why we should have a comment here because the error handling with the HOC is a little unintuitive. Should we then do this for all routes? If we do, should have a function that calls validateJson or something which contains this comment. If we do is it best placed in all routes or in the fetchPageData function. I'm just thinking about this because I have some unfinished work which kind of thinks about these things https://github.com/bbc/simorgh/pull/6852/files#diff-93bc8a6ae91fb1ee01233d35e3590c4d

ryanmccombe commented:
>Looks good to me. Just a question, is certain fields missing from data a common problem we have?

It's not common, but we have seen it happen for MAPs.  Because we just make a best attempt at rendering, those oddities were just going un-noticed at the time - so now we try to add some monitoring to the ARES fields we actually need

ryanmccombe committed:
Merge branch 'latest' into refactor-od-get-initial-data

ryanmccombe commented:
> I understand why we should have a comment here because the error handling with the HOC is a little unintuitive. Should we then do this for all routes? If we do, should have a function that calls validateJson or something which contains this comment. If we do is it best placed in all routes or in the fetchPageData function. I'm just thinking about this because I have some unfinished work which kind of thinks about these things https://github.com/bbc/simorgh/pull/6852/files#diff-93bc8a6ae91fb1ee01233d35e3590c4d

Aye - I left this pretty simple because I know you're working on a general approach.  Depending on how far we're going with this, I'd suggest considering not requiring each page define their own `getInitialData` function at all, but I'll post some ideas on that PR instead of here 😄

karinathomasbbc commented:
We had discussed it previously (and even added an issue for Live Radio https://github.com/bbc/simorgh/issues/6060) that if we didn't get any data to render for the H1, then we should display a 404 page (because that would be a P1 a11y issue). Would this help towards that? Also probably worth checking if this is even a requirement anymore.

ryanmccombe commented:
> We had discussed it previously (and even added an issue for Live Radio #6060) that if we didn't get any data to render for the H1, then we should display a 404 page (because that would be a P1 a11y issue). Would this help towards that? Also probably worth checking if this is even a requirement anymore.

This PR is just refactoring logging - PRs like this can turn into an endless cascade of niche / edge case behaviour handling, but I'm keen to just keep it scoped to the logging - there wasn't supposed to be any product changes here

I can't find this 404 in the requirements anyway - I can't imagine a 404 would be a better user experience than a page with no heading, but can ask product.
